### DISCUSSION TITLE
failed to create an iTwin project with error 422

### DISCUSSION THREAD

**Discussion by markgao085** [iTwin/itwinjs-core#4676] [Q&A]
trying to create an iTwin project with the following code, verified my access token along with header\body shown below in iTwin API 'try it out', it's working fine. but when running my code, it fails with error 422. what I'm missing here?

app.post('/project', function (req, res) {
    var projName = req.body.projectName;
    var projNo = req.body.projectNo;
    Axios({
        method: 'POST',
        url: 'https://api.bentley.com/projects',
        headers: {
            'Accept': 'application/vnd.bentley.itwin-platform.v1+json',
            'Content-Type': 'application/json',
            'Authorization': 'Bearer ' + access_token
        },
        data: querystring.stringify({
            "displayName": projName,
            "projectNumber": projNo,
            "geographicLocation": "Exton, PA, USA",
            "latitude": "40.032970",
            "longitude": "-75.631538",
            "timeZone": "Eastern Standard Time",
            "billingCountry": req.body.billingCountry,
            "status": "active",
            "dataCenterLocation": req.body.dataCenter,
            "allowExternalTeamMembers": true
        })
    })
        .then(function (response) {
            // Success
            console.log(response);
            res.send('iTwin project created successfully');
        })
        .catch(function (error) {
            // Failed
            console.log(error);
            res.send('Failed to create iTwin project');
        });
});

**Comment by HaleRazor73** [comment]
Hello @markgao085, could you post the response body for the 422 here (redacting sensitive information, if any, of course).

**Reply by markgao085** [reply]
here is the response

Error: Request failed with status code 422
    at createError (D:\BentleyRESTAPI\iTwinProjectPoC\node_modules\axios\lib\core\createError.js:16:15)
    at settle (D:\BentleyRESTAPI\iTwinProjectPoC\node_modules\axios\lib\core\settle.js:18:12)
    at IncomingMessage.handleStreamEnd (D:\BentleyRESTAPI\iTwinProjectPoC\node_modules\axios\lib\adapters\http.js:202:11)
    at IncomingMessage.emit (node:events:532:35)
    at endReadableNT (node:internal/streams/readable:1346:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  config: {
    adapter: [Function: httpAdapter],
    transformRequest: { '0': [Function: transformRequest] },
    transformResponse: { '0': [Function: transformResponse] },
    timeout: 0,
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    validateStatus: [Function: validateStatus],
    headers: {
      Accept: 'application/vnd.bentley.itwin-platform.v1+json',
      'Content-Type': 'application/json',
      Authorization: 'Bearer xxxx',
      'User-Agent': 'axios/0.18.1',
      'Content-Length': 279
    },
    method: 'post',
    url: 'https://api.bentley.com/projects',
    data: 'displayName=iTwinRESTAPIProjectPoc&projectNumber=iTwinRESTAPIProjectPoc&geographicLocation=Exton%2C%20PA%2C%20USA&latitude=40.032970&longitude=-75.631538&timeZone=Eastern%20Standard%20Time&billingCountry=US&status=active&dataCenterLocation=East%20US&allowExternalTeamMembers=true'
  },
  request: <ref *1> ClientRequest {
    _events: [Object: null prototype] {
      abort: [Function (anonymous)],
      aborted: [Function (anonymous)],
      error: [Function (anonymous)],
      socket: [Function (anonymous)],
      timeout: [Function (anonymous)],
      prefinish: [Function: requestOnPrefinish]
    },
    _eventsCount: 6,
    _maxListeners: undefined,
    outputData: [],
    outputSize: 0,
    writable: true,
    destroyed: false,
    _last: true,
    chunkedEncoding: false,
    shouldKeepAlive: false,
    maxRequestsOnConnectionReached: false,
    _defaultKeepAlive: true,
    useChunkedEncodingByDefault: true,
    sendDate: false,
    _removedConnection: false,
    _removedContLen: false,
    _removedTE: false,
    _contentLength: null,
    _hasBody: true,
    _trailer: '',
    finished: true,
    _headerSent: true,
    _closed: false,
    socket: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      secureConnecting: false,
      _SNICallback: null,
      servername: 'api.bentley.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'api.bentley.com',
      _readableState: [ReadableState],
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular *1],
      [Symbol(res)]: [TLSWrap],
      [Symbol(verified)]: true,
      [Symbol(pendingSession)]: null,
      [Symbol(async_id_symbol)]: 141,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: null,
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object],
      [Symbol(RequestTimeout)]: undefined
    },
    _header: 'POST /projects HTTP/1.1\r\n' +
      'Accept: application/vnd.bentley.itwin-platform.v1+json\r\n' +
      'Content-Type: application/json\r\n' +
      'Authorization: Bearer xxxx\r\n' +
      'User-Agent: axios/0.18.1\r\n' +
      'Content-Length: 279\r\n' +
      'Host: api.bentley.com\r\n' +
      'Connection: close\r\n' +
      '\r\n',
    _keepAliveTimeout: 0,
    _onPendingData: [Function: nop],
    agent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object: null prototype],
      requests: [Object: null prototype] {},
      sockets: [Object: null prototype],
      freeSockets: [Object: null prototype] {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      scheduling: 'lifo',
      maxTotalSockets: Infinity,
      totalSocketCount: 1,
      maxCachedSessions: 100,
      _sessionCache: [Object],
      [Symbol(kCapture)]: false
    },
    socketPath: undefined,
    method: 'POST',
    maxHeaderSize: undefined,
    insecureHTTPParser: undefined,
    path: '/projects',
    _ended: true,
    res: IncomingMessage {
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 3,
      _maxListeners: undefined,
      socket: [TLSSocket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      rawHeaders: [Array],
      rawTrailers: [],
      aborted: false,
      upgrade: false,
      url: '',
      method: null,
      statusCode: 422,
      statusMessage: 'Unprocessable Entity',
      client: [TLSSocket],
      _consuming: false,
      _dumped: false,
      req: [Circular *1],
      responseUrl: 'https://api.bentley.com/projects',
      redirects: [],
      [Symbol(kCapture)]: false,
      [Symbol(kHeaders)]: [Object],
      [Symbol(kHeadersCount)]: 14,
      [Symbol(kTrailers)]: null,
      [Symbol(kTrailersCount)]: 0,
      [Symbol(RequestTimeout)]: undefined
    },
    aborted: false,
    timeoutCb: null,
    upgradeOrConnect: false,
    parser: null,
    maxHeadersCount: null,
    reusedSocket: false,
    host: 'api.bentley.com',
    protocol: 'https:',
    _redirectable: Writable {
      _writableState: [WritableState],
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      _options: [Object],
      _redirectCount: 0,
      _redirects: [],
      _requestBodyLength: 279,
      _requestBodyBuffers: [],
      _onNativeResponse: [Function (anonymous)],
      _currentRequest: [Circular *1],
      _currentUrl: 'https://api.bentley.com/projects',
      [Symbol(kCapture)]: false
    },
    [Symbol(kCapture)]: false,
    [Symbol(kNeedDrain)]: false,
    [Symbol(corked)]: 0,
    [Symbol(kOutHeaders)]: [Object: null prototype] {
      accept: [Array],
      'content-type': [Array],
      authorization: [Array],
      'user-agent': [Array],
      'content-length': [Array],
      host: [Array]
    }
  },
  response: {
    status: 422,
    statusText: 'Unprocessable Entity',
    headers: {
      'content-length': '113',
      'request-context': 'appId=cid-v1:caffab81-b80a-40d1-a9e3-3d1a07228483',
      'x-correlation-id': 'dc870436-6e89-4c15-a5b0-f01edab1e3ef',
      'x-itwinplatform-media-type': 'bentley.itwin-platform.v1',
      'x-itwinplatform-region': 'East US',
      date: 'Thu, 17 Nov 2022 18:06:51 GMT',
      connection: 'close'
    },
    config: {
      adapter: [Function: httpAdapter],
      transformRequest: [Object],
      transformResponse: [Object],
      timeout: 0,
      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxContentLength: -1,
      validateStatus: [Function: validateStatus],
      headers: [Object],
      method: 'post',
      url: 'https://api.bentley.com/projects',
      data: 'displayName=iTwinRESTAPIProjectPoc&projectNumber=iTwinRESTAPIProjectPoc&geographicLocation=Exton%2C%20PA%2C%20USA&latitude=40.032970&longitude=-75.631538&timeZone=Eastern%20Standard%20Time&billingCountry=US&status=active&dataCenterLocation=East%20US&allowExternalTeamMembers=true'
    },
    request: <ref *1> ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 6,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      destroyed: false,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      maxRequestsOnConnectionReached: false,
      _defaultKeepAlive: true,
      useChunkedEncodingByDefault: true,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: null,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      _closed: false,
      socket: [TLSSocket],
      _header: 'POST /projects HTTP/1.1\r\n' +
        'Accept: application/vnd.bentley.itwin-platform.v1+json\r\n' +
        'Content-Type: application/json\r\n' +
        'Authorization: Bearer xxxxx\r\n' +
        'User-Agent: axios/0.18.1\r\n' +
        'Content-Length: 279\r\n' +
        'Host: api.bentley.com\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _keepAliveTimeout: 0,
      _onPendingData: [Function: nop],
      agent: [Agent],
      socketPath: undefined,
      method: 'POST',
      maxHeaderSize: undefined,
      insecureHTTPParser: undefined,
      path: '/projects',
      _ended: true,
      res: [IncomingMessage],
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      host: 'api.bentley.com',
      protocol: 'https:',
      _redirectable: [Writable],
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype]
    },
    data: { error: [Object] }
  }
}

**Reply by davidhjones** [reply]
What will be helpful is the `data` object you see in the response here. The `error` json returned in the response to your request has very valuable information to explain what is going wrong. Can you post the actual json response you are getting while making the request (rather than this axios javascript object)?

**Reply by davidhjones** [reply]
if you inspect your `response.data.error` json object and print the contents, that should be what we're looking for. You should see something like:
```
{
    "error": {
        "code": "InvalidCreateProjectRequest",
        "message": "Cannot create project.",
        "details": [{
                "code": "MissingRequiredProperty",
                "message": "A property required to create a project is missing.",
                "target": "displayName"
            }
        ]
    }
}
```

**Reply by markgao085** [reply]
I'm new to this, do you mind showing me how to get the response.data.error json object? I tried with the following, but didn't work out.

.then(function (response) {
            // Success
            console.log(response);
            res.send('iTwin project created successfully');
        })
        .catch(function (error) {
            // Failed
            console.log(res.json());
            //console.log(error);
            res.send('Failed to create iTwin project');
        });

**Reply by davidhjones** [reply]
How did you print that json above? Which line of code was that?

**Reply by markgao085** [reply]
console.log(res.json());

**Reply by davidhjones** [reply]
In the json you sent you’ll see a property `data` (close to the very bottom). You’ll want to access this property. Try `console.log(res.data.error)`

**Comment by davidhjones** [comment]
@markgao085 -- I would also suggest that you look into using our latest API -- the [iTwins API](https://developer.bentley.com/apis/itwins/). It supersedes the Projects API. From the sounds of it you're in early stages of development, so again I would suggest that begin by using the latest and greatest APIs. 

As @KevinHale-Bentley suggested -- providing deatils of the 422 response would be helpful in diagnosing the problem at hand.

**Reply by markgao085** [reply]
thanks, will give it a try

**Reply by markgao085** [reply]
with the iTwins API, I managed to create an iTwin project successfully, but it turns out no option to create it on https://projectwise365.bentley.com/, any plan to accommodate which portal to use?
